### ISSUE TITLE
Error with source maps & jsdom

### ISSUE THREAD

**Issue by thom-nic** [zinserjan/mocha-webpack#54]
Using mocha-babel with jsdom seems to causes some confusion with source-map-support...

```
$ $(npm bin)/mocha-webpack --opts test/mocha-webpack.opts
/Users/thom/dev/sandbox/app-client/node_modules/source-map-support/source-map-support.js:411
      var hasStack = (arguments[1] && arguments[1].stack);
                                                  ^

TypeError: Invalid URL
```

My webpack test config, pretty much straight from the docs:

``` javascript
const nodeExternals = require('webpack-node-externals');

module.exports = {
  target: 'node', // in order to ignore built-in modules like path, fs, etc.
  externals: [nodeExternals()], // in order to ignore all modules in node_modules folder
  module: { .... }
  output: {
    // sourcemap support for IntelliJ/Webstorm
    devtoolModuleFilenameTemplate: '[absolute-resource-path]',
    devtoolFallbackModuleFilenameTemplate: '[absolute-resource-path]?[hash]'
  },
  devtool: 'cheap-module-source-map',
};
```

mocha-webpack.opts:

```
--recursive
--colors
--reporter list
--ui bdd
--timeout 2000
--require source-map-support/register
--require test/register.js
--webpack-config config/webpack.test.config.js
**/__tests__/*.js
```

test/register.js sets up jsdom:

``` javascript
//require('babel-register')({
  // make sure you use the same presets as in webpack.config.js:
//  presets: ['react', 'es2015', 'stage-1'],
//});
const jsdom = require('jsdom').jsdom;

const exposedProperties = ['window', 'navigator', 'document'];

const document = jsdom('');
global.document = document;
global.window = document.defaultView;
Object.keys(document.defaultView).forEach((property) => {
  if (typeof global[property] === 'undefined') {
    exposedProperties.push(property);
    global[property] = document.defaultView[property];
  }
});

global.navigator = {
  userAgent: 'node.js',
};
```

I've tried adding `require('source-map-support/register');` at the top or bottom of register.js and also tried changing the order of `--register` lines in my mocha-webpack.opts with no success...

**Comment by thom-nic** [comment]
Ok I think I figured out, instead of using `require('source-map-support/register');` I needed to do the following, I put it at the bottom of `register.js` after I setup jsdom:

```
require('source-map-support').install({environment:'node'});
```

This appears to give an accurate sourcemap if an exception is thrown from my application code.  However it doesn't seem to provide correct source maps for the actual test files:

test example.js:
```
describe('a failing test', () => {
  it("should print a stack trace pointing to this file", () => {
    throw new Error('boom');
  });
});
```

When I run it.... (note the last line)
```
$ $(npm bin)/mocha-webpack --opts test/mocha-webpack.opts

  1) a failing test should print a stack trace pointing to this file

  0 passing (11ms)
  1 failing

  1) a failing test should print a stack trace pointing to this file:
     Error: boom
      at Context.<anonymous> (.tmp/mocha-webpack/09e4ba4225623e5575afb37f2f258d2f:11:1)
```

**Comment by lifehackett** [comment]
So replacing `cheap-module-source-map` with `source-map` fixed it for me.

**package.json**
```
"test": "mocha-webpack --webpack-config webpack-test.config.js --glob '*.test.js' test --require ./test/test-helper.js --recursive"
```


**test-helper.js**
```
var chai = require('chai');
var sinonChai = require('sinon-chai');
var jsdom = require('jsdom').jsdom;

chai.use(sinonChai);


global.document = jsdom('');
global.window = document.defaultView;
Object.keys(document.defaultView).forEach((property) => {
  if (typeof global[property] === 'undefined') {
    global[property] = document.defaultView[property];
  }
});

global.navigator = {
  userAgent: 'node.js'
};


require('source-map-support').install({
  handleUncaughtExceptions: false,
  environment: 'node',
});
```


**webpack-test.config.js**
```
var webpack = require('webpack');
var path = require('path');
var nodeExternals = require('webpack-node-externals');

module.exports = {
  devtool: 'source-map',
  target: 'node',
  externals: [nodeExternals()],
  output: {
    // sourcemap support for IntelliJ/Webstorm
    devtoolModuleFilenameTemplate: '[absolute-resource-path]',
    devtoolFallbackModuleFilenameTemplate: '[absolute-resource-path]?[hash]'
  },
  module: {
    loaders: [{
      test: /\jsx?$/,
      loader: 'babel',
      include: [
        path.join(__dirname, 'src'),
        path.join(__dirname, 'test'),
      ],
      exclude: /node_modules/,
      query: {
        presets: ['es2015', 'react'],
        plugins: ['transform-object-rest-spread', 'jsx-control-statements']
      }
    }]
  },
  resolve: {
    root: path.resolve(__dirname),
    alias: {
      components: 'src/components',
      enums: 'src/util/enums',
      foo: 'src/util',
    },
    extensions: ['', '.js']
  }
};
```

**Comment by thom-nic** [comment]
You're right, using `devtool: 'source-map'` fixes it. (I also tried `cheap-source-map` which gave the correct file name but not the correct line number...)

Also interesting to note is if I specified a specific file (not directory) to test the results differ:
```
± $(npm bin)/mocha-webpack --opts test/mocha-webpack.opts src/__tests__/appTest.jsx 

  ․ <App /> has a header: 10ms
  1) <App /> calls componentDidMount

  1 passing (44ms)
  1 failing

  1) <App /> calls componentDidMount:
     Error: boom
      at Context.<anonymous> (.tmp/mocha-webpack/appTest.jsx:80:12)
```
This however isn't the worst thing since that points to the correct *compiled* line that threw the error.  Anyway, thanks that tip provides a nice workable solution!

**Comment by zinserjan** [comment]
Does the`cheap-module-source-map` issue only occur in conjunction with jsdom?

**Comment by kristianmandrup** [comment]
I'm having the same problem. I've been trying with and without the `source-map`. Haven't tried the `cheap` variant yet. 

```
// webpack.test.config.js
module.exports = {
  devtool: 'source-map',
  target: 'node', // in order to ignore built-in modules like path, fs, etc.
  externals: [nodeExternals()], // in order to ignore all modules in node_modules folder
  module: {
    rules: [
      {
        enforce: 'pre',
        test: /\.jsx?$/,
        exclude: /node_modules/,
        loader: 'eslint'
      },
      {
        test: /\.jsx?$/,
        exclude: /node_modules/,
        loader: 'babel',
        query: {
          babelrc: true,
          cacheDirectory: true
        }
      }
    ]
  }
}
```

`mocha-webpack.opts`

```
--bail
--check-leaks
--reporter spec
--ui bdd
--colors
--require src/helpers/index.js
--webpack-config build/webpack.test.config.js
src/{**/,/}*.test.js
```

My `helpers/index.js`

```js
require('source-map-support').install({
  handleUncaughtExceptions: false,
  environment: 'node'
})
```

`package.json`

```js
  "scripts": {
    "mocha:test": "npm run lint & mocha-webpack",
     "lint": "eslint src/**/*.js",
}
```

Error:

```
webpack test env { devtool: 'source-map',
  target: 'node',
  externals: [ [Function] ],
  module: { rules: [ [Object], [Object] ] } }



  Operator: code
    an instance
      1) have a name


  0 passing (8ms)
  1 failing

  1) Operator: code an instance have a name:
     TypeError: (0 , _chai.expect)(...).to.be is not a function
      at Context.that.should (.tmp/mocha-webpack/417cb2282a69273b762633c8038abca7/webpack:/Users/kristianmandrup/repos/node-libs/code-operator/src/operator/code.test.js:9:23)
```

Any ideas? Looks like the source mapping back to the original code doesn't work?

**Comment by kristianmandrup** [comment]
So many source map options... 

http://cheng.logdown.com/posts/2016/03/25/679045
http://survivejs.com/webpack/developing-with-webpack/enabling-sourcemaps/

`eval-source-map` and `cheap-eval-source-map` are recommended for dev

However I can't seem to find any of these source map modules on npm now? what gives!?

**Comment by kristianmandrup** [comment]
Full list of [webpack source map options]( https://webpack.github.io/docs/configuration.html#devtool)

Trying to run tests which use `import` and all the other "good stuff".
Haven't yet enabled the [webpack source loader](https://github.com/webpack/source-map-loader)

```js
import { expect } from 'chai'
import code from './code'
import test from 'mocha-test-dsl'

test('Operator: code')
  .that('an instance')
  .should('have a name', () => {
    let result = code('kris')
    expect(result).to.be(true)
  })
  .run()
```

**Comment by kristianmandrup** [comment]
I've also seen this being used `mocha --compilers js:babel-core/register` 

I think I recall that is how I made mocha work with babel before I started trying to add webpack to the mix. Why is the `--compilers` option not supported by `mocha-webpack` then?

**Comment by kristianmandrup** [comment]
I tried with `mocha src/**/*.test.js --compilers js:babel-core/register` but I get exactly the same problem/error. Source map issue for sure :(

**Comment by kristianmandrup** [comment]
WTF, trying these simple guides for babel ES6 with mocha. 
Still same error ;(

http://x-team.com/2016/05/setting-up-javascript-testing-tools-for-es6/
http://krasimirtsonev.com/blog/article/using-mocha-with-es6-spec-files

`mocha --compilers js:babel-register --require babel-polyfill` doesn't even work!? 

AAAARRGGHH!

**Comment by kristianmandrup** [comment]
Good damn, it works! Simply too tired to see and think clearly... struggling for hours plays tricks with your mind! 

The error was bad expect syntax.. really `to.be` is not valid syntax, fixed with `to.eql()`. Yikes!!!

**Comment by kristianmandrup** [comment]
To be clear, I've now tested the following `devtool` options and they all work

```
  devtool: 'source-map'
  devtool: 'cheap-module-eval-source-map'
  devtool: 'cheap-eval-source-map'
  devtool: 'eval-source-map'
  devtool: 'cheap-module-source-map'
```

Even works for me without the `devtool` set. Hurray!

**Comment by miadwang** [comment]
I am having a similar problem. I have tried 'source-map' and 'cheap-module-source-map'. Both didn't work.

```
// package.json
"scripts": {
    "test": "BABEL_ENV=test mocha-webpack --require mocha-setup.js --webpack-config webpack.config-test.js --watch 'src/components/**/*.spec.js'"
  }
```

```
// mocha-setup.js

require('source-map-support').install({
  handleUncaughtExceptions: false,
  environment: 'node',
});
```

```
// webpack-config-test.js

var nodeExternals = require('webpack-node-externals');

module.exports = {
  target: 'node',
  externals: [nodeExternals()],
  devtool: 'source-map',
  resolve: {
    extensions: ['', '.js', '.jsx']
  },
  output: {
    devtoolModuleFilenameTemplate: '[absolute-resource-path]',
    devtoolFallbackModuleFilenameTemplate: '[absolute-resource-path]?[hash]'
  },
  module: {
    loaders: [
      { test: /\.jsx?$/, loader: 'babel-loader', exclude: /node_modules/ },
      { test: /\.css$/, loaders: [
        'css-loader/locals?importLoaders=1&modules&localIdentName=[local]',
      ], exclude: /node_modules/ },
      { test: /\.scss$/, loaders: [
        'css-loader/locals?importLoaders=1&modules&localIdentName=[local]',
        'sass-loader'
      ], exclude: /node_modules/ }
    ]
  }
};
```

```
// Button.spec.js

import React from 'react';
import { expect } from 'chai';
import { shallow } from 'enzyme';
import Button from './Button';

describe('<Button />', () => {
  it('should render properly', () => {
    const wrapper1 = shallow(
      <Button
        text="Test"
      />
    );
    expect(wrapper1.find('button-error')).to.have.length(1);
});
```

```
// error

<Button />
    1) should render properly
    ✓ should react to actions


  1 passing (4ms)
  1 failing

  1) <Button /> should render properly:
     AssertionError: expected { Object (root, unrendered, ...) } to have a length of 1 but got 0
      at Context.<anonymous> (.tmp/mocha-webpack/01626fa44c1893dd7d71f43c1bac677d/01626fa44c1893dd7d71f43c1bac677d-output.js:112:63)
```

What did I miss?

**Comment by miadwang** [comment]
If the error was inside expect(), the source map is correct.

```
// Button.spec.js

import React from 'react';
import { expect } from 'chai';
import { shallow } from 'enzyme';
import Button from './Button';

describe('<Button />', () => {
  it('should render properly', () => {
    const wrapper1 = shallow(
      <Button
        text="Test"
      />
    );
    // expect(wrapper1.find('button-error')).to.have.length(1);
    expect(wrapper1.find('button-error').text()).to.equal('Test');
});
```

```
// error

 <Button />
    1) should render properly
    ✓ should react to actions


  1 passing (4ms)
  1 failing

  1) <Button /> should render properly:
     Error: Method “text” is only meant to be run on a single node. 0 found instead.
      at ShallowWrapper.single (node_modules/enzyme/build/ShallowWrapper.js:1436:17)
      at ShallowWrapper.text (node_modules/enzyme/build/ShallowWrapper.js:715:21)
      at Context.<anonymous> (src/components/Button/Button.spec.js:19:7)
```

It seems that source map works differently for enzyme shallow and chai expect.

**Comment by zinserjan** [comment]
Sourcemaps works out of the box with v1.0.0. For configuration have a look at http://zinserjan.github.io/mocha-webpack/docs/installation/webpack-configuration.html#sourcemaps


There were also several sourcemap issues in webpack that were fixed with webpack@2.3.2.
